syntax = "proto3";
package ledger;

import "google/protobuf/timestamp.proto";

message Posting {
    double amount = 1;
    string side = 2;
    string account_id = 3;
}

message PostEntryRequest {
    string payee = 1;
    string note = 2;
    string external_id = 3;
    string category_id = 4;
    google.protobuf.Timestamp posted_at = 5;
    repeated Posting postings = 6;
    string account_id = 7; // account for tx
    string user_id = 8;
}

message Txn {
    string id = 1;
    google.protobuf.Timestamp posted_at = 2;
    string payee = 3;
    string note = 4;
    string external_id = 5;
    string category_id = 6;
    string user_id = 7;
}

message TxnId {
    string id = 1;
}

message BalanceRequest {
    string account_id = 1;
    google.protobuf.Timestamp at = 2;
}

message BalanceResponse {
    double amount = 1;
}

message StreamRequest {
    string account_id = 1;
    google.protobuf.Timestamp start = 2;
    google.protobuf.Timestamp end = 3;
}

service LedgerService {
    rpc PostEntry(PostEntryRequest) returns (TxnId);
    rpc GetBalance(BalanceRequest) returns (BalanceResponse);
    rpc StreamTxns(StreamRequest) returns (stream Txn);
}

